### PR TITLE
School Library: Preserve Data

### PR DESCRIPTION
**Pull request by mateo951**
- Added method to preserve data
- Added save data for books and people
- Added fix for students
- Added method to read teachers and students
- Fixed for preserving data v1
- Added fix for linters
- Added method for rentals

### CODE REVIEW

**Review by mke2111 — CHANGES_REQUESTED** [review]
### Hey @mateo951, @JbirdL86 & @Felipeg005, 

**This is a great project you have going on, well-done lads**. 

   
**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

## Required Changes ♻️
_Check the inline comments under the review for sections to improve._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: library.rb [hunk 1]
```diff
@@ -49,4 +52,102 @@ def list_rentals
...
+    JSON.generate(book_data)
+  end
+
+  def parse_rental_json(data_array)
+    rental_data = []
+    data_array.each do |data|
+      rental_data.push({ Date: data.date, BookId: data.book.id, PersonId: data.person.id })
+    end
+    JSON.generate(rental_data)
+  end
```
Comment (mke2111) [inline library.rb:94]: - [x] On exit from the app, the data should be saved (all the data: books, people, and rentals) into JSON files. One per type of data should be stored and named accordingly (e.g. rentals.json). Currently, all these json files are missing.
To achieve this functionality, add something like this when a user exits the app.
```
File.open("rentals.json", "w") do |f|
      f.write JSON.generate(@rentals)
    end
```

- [x] You need to save all the data i.e. people, books and rentals.

File: library.rb [hunk 2]
```diff
@@ -49,4 +52,102 @@ def list_rentals
...
+    file_data = file.read if file
+    rentals_data = JSON.parse(file_data)
+    rentals = []
+    rentals_data.each do |data|
+      rental_book = @book_list.find { |book| book.id == data['BookId'] }
+      rental_person = @people.find { |person| person.id == data['PersonId'] }
+      rentals.push(Rental.new(data['Date'], rental_book, rental_person))
+    end
+    @rentals = rentals
+  end
```
Comment (mke2111) [inline library.rb:140]: - [x] On startup of the app, load all data from files. Currently, previous data is not preserved. Check the image below. A student is added but the list is returned empty.
Screenshot from 2021-11-25 09-14-40

File: library.rb (hunk 1 above)
Comment (Felipeg005) [reply library.rb:94]: Thanks for the review, we had a bug and now it's fixed 👍

File: library.rb (hunk 2 above)
Comment (Felipeg005) [reply library.rb:140]: Thank you mke2111, now this bug 🐛 it's fixed 😄

Comment (DeliceLydia) [reply library.rb:140]: Solved this!

File: library.rb (hunk 1 above)
Comment (DeliceLydia) [reply library.rb:94]: Well resolved!

**Review by DeliceLydia — APPROVED** [review]
# Congratulations! 🎉 

## Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



### Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard).. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**